worlds/witness/player_logic.py

```diff
@@ -794,23 +846,124 @@ def determine_unrequired_entities(self, world: "WitnessWorld") -> None:
+                # However, I don't think they should be hunt panels in this case.
+                world.options.disable_non_randomized_puzzles
+                and not world.options.shuffle_discarded_panels
+                and entity_obj["locationType"] == "Discard"
+            )
+            and entity_hex not in disallowed_entities_for_panel_hunt
+            and entity_hex not in self.HUNT_ENTITIES
+        ]
+
+        total_panels = world.options.panel_hunt_total.value
+        same_area_discouragement_factor = world.options.panel_hunt_discourage_same_area_factor / 100
+
+        # If we're using random picking, just choose all the panels now and return
+        if not same_area_discouragement_factor:
+            hunt_panels = world.random.choices(eligible_panels, k=total_panels - len(self.HUNT_ENTITIES))
+            self.HUNT_ENTITIES.update(hunt_panels)
+            return
+
+        # Make a lookup of eligible hunt panels per area
+        eligible_panels_by_area = defaultdict(set)
+        for eligible_panel in eligible_panels:
+            associated_area = static_witness_logic.ENTITIES_BY_HEX[eligible_panel]["area"]["name"]
+            eligible_panels_by_area[associated_area].add(eligible_panel)
+
+        # If we're discouraging panels from the same area being picked, we have to pick panels one at a time
+        # For higher total counts, we do them in small batches for performance
+        batch_size = max(1, total_panels // 20)
+
+        contributing_percentage_per_area = {area: 0 for area in eligible_panels_by_area}
 
-        name = self.REFERENCE_LOGIC.ENTITIES_BY_HEX[entity_hex]["checkName"] + action
-        if entity_hex not in self.USED_EVENT_NAMES_BY_HEX:
-            warning(f'Entity "{name}" does not have an associated event name.')
-            self.USED_EVENT_NAMES_BY_HEX[entity_hex] = name + " Event"
-        pair = (name, self.USED_EVENT_NAMES_BY_HEX[entity_hex])
-        return pair
+        while len(self.HUNT_ENTITIES) < total_panels:
+            actual_amount_to_pick = min(batch_size, total_panels - len(self.HUNT_ENTITIES))
+
+            max_percentage = max(contributing_percentage_per_area.values())
+            if max_percentage == 0:
+                allowance_per_area = {area: 1 for area in contributing_percentage_per_area}
+            else:
+                allowance_per_area = {
+                    area: (max_percentage - current_percentage) / max_percentage
+                    for area, current_percentage in contributing_percentage_per_area.items()
+                }
+                # use same_area_discouragement as lerp factor
+                allowance_per_area = {
+                    area: (1.0 - same_area_discouragement_factor) + (weight * same_area_discouragement_factor)
+                    for area, weight in allowance_per_area.items()
+                }
+
+            assert min(allowance_per_area.values()) >= 0, (f"Somehow, an area had a negative weight when picking"
```

Comment:
Do you want this to be an `assert` or something like an exception so it'll still error when built?

Reply:
I prefer the assert here, this is more for me to notice that it happened than to crash generation - A negative weight is treated as 0, so I'd be crashing a generation that would be successful otherwise